include/swift/AST/DiagnosticsSema.def

```diff
@@ -365,6 +365,35 @@ ERROR(cannot_convert_argument_value_generic,none,
       "cannot convert value of type %0 (%1) to expected argument type %2 (%3)",
       (Type, StringRef, Type, StringRef))
 
+ERROR(cannot_convert_non_ephemeral,none,
+      "temporary pointer argument of type %0 cannot be passed to parameter "
+      "expecting pointer that outlives the duration of the call", (Type))
+WARNING(cannot_convert_non_ephemeral_warning,none,
+        "passing temporary pointer argument of type %0 to parameter expecting "
+        "pointer that outlives the duration of the call leads to undefined "
+        "behaviour; this will be an error in a future release", (Type))
```

Comment:
This diagnostic feels a bit too wordy, but I can't think of how to improve it – any ideas @jrose-apple?

Reply:
Hm, well, we could remove that "u" from "behavior". :-)

> warning: temporary pointer will be destroyed after call to %0

Okay, that sounds sufficiently scary, but the problem is that the user probably doesn't think they've written a pointer. They wrote an inout expression or an array or string.

> error: cannot pass array to %0; argument %1 must be a pointer that outlives the call to %0
> error: cannot pass string to %0; argument %1 must be a pointer that outlives the call to %0
> error: cannot use inout expression here; argument %1 must be a pointer that outlives the call to %0

How does that sound? And then we get the warning form for pre-Swift-5 (which we maybe take away completely in the future):

> warning: passing array to %0, but argument %1 should be a pointer that outlives the call to %0
> warning: passing string to %0, but argument %1 should be a pointer that outlives the call to %0
> warning: inout expression creates a temporary pointer, but argument %1 should be a pointer that outlives the call to %0

I don't love this but it's a little less wordy?

Reply:
Oops, need to start using a US spell checker :) Thanks for the suggestions, I like them!

IMO it would be nice to use the argument type of the array or string in the diagnostic message, and avoid repeating the name of the callee – so how about the following?

> error: cannot pass %0 to parameter; argument must a pointer that outlives the call%select{| to %2}1

> warning: passing %0 to parameter, but argument should be a pointer that outlives the call%select{| to %2}1; this will be an error in a future release

> error: cannot use inout expression here; argument must a pointer that outlives the call%select{| to %1}0

> warning: inout expression creates a temporary pointer, but argument should be a pointer that outlives the call%select{| to %1}0; this will be an error in a future release

For now, I've omitted the placeholder for "argument %1", as I wasn't too sure what it would be filled in with – would it be the argument position or the text of the argument expr?

I've gone ahead and pushed a commit with the message changes so far.

Reply:
Typo: "must a pointer".

The reason I had the name of the callee twice was that "to parameter" on its own sounds weird to me. The "argument %1" part was the position in my head, to make it slightly easier to tell what the diagnostic is referring to when just looking at the line and not the column or highlight.

I feel like the "this will be an error in a future release" isn't really adding that much value either, especially for a message that's already fairly long.

Reply:
> Typo: "must a pointer"

Oops, good catch!

> The reason I had the name of the callee twice was that "to parameter" on its own sounds weird to me.

Interesting! "to parameter" sounds natural to me – nonetheless I'm happy to change it. Though for cases where we don't have a callee decl, should it be "to parameter" or "to call"? The former sounds more natural to me, but the latter is more consistent with the message when we do have a callee.

> The "argument %1" part was the position in my head, to make it slightly easier to tell what the diagnostic is referring to when just looking at the line and not the column or highlight.

Thanks for clarifying, I'll go ahead and make that change.

> I feel like the "this will be an error in a future release" isn't really adding that much value either, especially for a message that's already fairly long.

Fair enough – I added it because I felt the diagnostic wasn't quite scary enough without it, but given the length of the message, I agree it probably doesn't quite pull its weight.